sklearn/cluster/hierarchical.py

```diff
@@ -80,6 +80,48 @@ def _fix_connectivity(X, connectivity, affinity):
     return connectivity, n_components
 
 
+def _single_linkage_tree(connectivity, n_samples, n_nodes, n_clusters,
+                         n_components, return_distance):
+    """
+    Perform single linkage clustering on sparse data via the minimum
+    spanning tree from scipy.sparse.csgraph, then using union-find to label.
+    The parent array is then generated by walking through the tree.
+    """
+    from scipy.sparse.csgraph import minimum_spanning_tree
+
+    # Ensure zero distances aren't ignored by setting them to "epsilon"
+    epsilon_value = connectivity.data[connectivity.data > 0].min() * 1E-8
+    connectivity.data[connectivity.data == 0] = epsilon_value
+
+    # Use scipy.sparse.csgraph to generate a minimum spanning tree
+    mst = minimum_spanning_tree(connectivity.tocsr())
+
+    # Convert the graph to scipy.cluster.hierarchy array format
+    mst = mst.tocoo()
```

Comment:
At this point, we probably need to undo the epsilon values, e.g.

```python
mst.data[mst.data == epsilon_value] = 0
```

Relatedly, I have some concerns about the epsilon approach overall: for example, it will not work in the case that the smallest value of the array is within a factor of 1E8 of the smallest representable value. We could address that by explicitly setting the epsilon value to the smallest representable positive value:
```
epsilon = np.nextafter(0, 1, dtype='float64')
```
(assuming we're certain that the dtype is double) then the only problematic case appears if the array already contains this value, at which point it's close enough to zero that it probably doesn't matter in any case of interest.

Reply:
By the way, does anyone know if cython or scipy.sparse would have issues with subnormal floating point numbers?

Reply:
Yes, that makes reasonable sense. Would it be better to use the dtype of the connectivity matrix to be safe in case of float32? Although that, in turn, will raise serious issues in the case of an integer valued connectivity matrix (which could theoretically occur for some metrics I suppose). One option would be to force case to float64 under the assumption that it is a sparse matrix and memory is "cheap". I would be interested in your opinion on the best option.

Reply:
Yeah, using the matrix dtype makes sense.

 If the connectivity matrix is integer, even the code as currently written will fail because epsilon will be cast back to zero. Is there any reason to expect integer connectivities? I thought this function was only called on the output of a distance computation.

Reply:
It should only contain the output of distance computations, I was just being extra cautious at this point. It looks like ``paired_distances`` is guaranteed to return float64 by the way the result is initialised. If the user passes in affinity as ``'precomputed'`` then connectivity will get populated by values in ``X``. If they call ``linkage_tree`` directly then there is no dtype checking done that I can see to ensure the values are floating point. Adding a cast to the precomputed case makes reasonable sense, however, and should obviate the issue.

Reply:
I meant to add here: nternally the ``minimum_spanning_tree`` function [converts all inputs to float64](https://github.com/scipy/scipy/blob/e744034d37f63cdf98ceef30511fd4c16ad2d500/scipy/sparse/csgraph/_min_spanning_tree.pyx#L85), so doing the conversion here won't incur any cost.

Reply:
That makes life easier then; thanks for pointing it out, I hadn't thought to check there.